src/collect.jl

```diff
@@ -0,0 +1,87 @@
+collectcolumns(itr) = collectcolumns(itr, Base.iteratorsize(itr))
+
+function collectcolumns(itr, ::Union{Base.HasShape, Base.HasLength})
+    st = start(itr)
+    el, st = next(itr, st)
+    dest = similar(arrayof(typeof(el)), length(itr))
+    dest[1] = el
+    collect_to_columns!(dest, itr, 2, st)
+end
+
+function collect_to_columns!(dest::AbstractArray{T}, itr, offs, st) where {T}
+    # collect to dest array, checking the type of each result. if a result does not
+    # match, widen the result type and re-dispatch.
+    i = offs
+    while !done(itr, st)
+        el, st = next(itr, st)
+        if fieldwise_isa(el, T)
+            @inbounds dest[i] = el
+            i += 1
+        else
+            new = widencolumns(dest, i, el, T)
+            @inbounds new[i] = el
+            return collect_to_columns!(new, itr, i+1, st)
+        end
+    end
+    return dest
+end
+
+function collectcolumns(itr, ::Base.SizeUnknown)
+    st = start(itr)
+    el, st = next(itr, st)
+    dest = similar(arrayof(typeof(el)), 1)
+    dest[1] = el
+    grow_to_columns!(dest, itr, 2, st)
+end
+
+function grow_to_columns!(dest::AbstractArray{T}, itr, offs, st) where {T}
+    # collect to dest array, checking the type of each result. if a result does not
+    # match, widen the result type and re-dispatch.
+    i = offs
+    while !done(itr, st)
+        el, st = next(itr, st)
+        if fieldwise_isa(el, T)
+            push!(dest, el)
+            i += 1
+        else
+            new = widencolumns(dest, i, el, T)
+            push!(new, el)
+            return grow_to_columns!(new, itr, i+1, st)
+        end
+    end
+    return dest
+end
+
+@generated function fieldwise_isa(el::S, ::Type{T}) where {S<:Tup, T}
+    if all((s <: t) for (s, t) in zip(S.parameters, T.parameters))
+        return :(true)
+    else
+        return :(false)
+    end
+end
+
+@generated function fieldwise_isa(el::S, ::Type{T}) where {S, T}
+    if S <: T
+        return :(true)
+    else
+        return :(false)
+    end
+end
```

Comment:
I think this doesn't need to be `@generated`, a `Base.@pure` or `@inline` might work here. But I guess once a package has `@generated` it really doesn't matter how many of them are there :-p

Reply:
`Base.@pure` seemed to not work as nicely, maybe we can leave as is now and do a "remove all generated functions" PR in the future (maybe in Julia 0.7).